cmd/influxd/run/server.go

```diff
@@ -65,6 +65,12 @@ type BuildInfo struct {
 	Time    string
 }
 
+type StartupProgress interface {
+	AddShard()
+	RemoveShardFromCount()
+	CompletedShard()
+}
```

Comment:
I understand the distinction between `RemoveShardFromCount` and `CompleteShard`, but I'm not sure it makes sense as a user watches the startup to see the total number of shards go down. Especially since if they do a quick `ls` on the disk they will disagree about how many shards there.  I would just keep `CompletedShard`. If you wanted to show how many succeeded vs failed, you could add a bool parameter to keep count of succeeded vs failed shard opens.

Also, if you don't use this interface, don't include it. It's helpful to be able to think about it conceptually, but if you don't actually use it then it is dead code.

Reply:
Although it looks like you do actually use it. If so, never mind that part of the comment.

Reply:
I am using it to correct the shard count on the fly - right now the logic is that if a shard fails to get opened it gets removed from the total. I could technically call `CompletedShard()` upon failure as well? In theory it's completed. That way I could also cut out the check https://github.com/influxdata/influxdb/pull/25378/files/2426bc876bd25503bc5f92a8447eb4a636a34c4b#diff-cdfd2e18a0a510f627c274bb98ad42cabfdedc05f5434da4fcf6fe4e629aa989R27 here as well to remove the race condition.